File: device-sdk/src/main/java/io/oddworks/device/model/Metric.java
package io.oddworks.device.model;

import java.util.Map;

import io.oddworks.device.Util;

public class Metric {
    private final String type;
    private final Map<String, Object> attributes;
    public static final String ENABLED = "enabled";
    public static final String INTERVAL = "interval";
    public static final String ACTION = "action";


    public Metric(String type, Map<String, Object> attributes) {
        this.type = type;
        this.attributes = attributes;
    }

    public String getType() {
        return type;
    }

    @SuppressWarnings("unchecked")
    public <T> T get(String key) {
        switch (key) {
            case ENABLED:
                Boolean enabled = (Boolean) attributes.get(key);
                if (enabled == null) {
                    return (T) Boolean.FALSE;
                }
                return (T) enabled;
            case INTERVAL:
                return (T) (Util.getInteger(attributes, key, null));
            default:
                return (T) (Util.getString(attributes, key, null));
        }
    }

    public Map<String, Object> getAttributes() {
        return attributes;
    }
}

File: device-sdk/src/main/java/io/oddworks/device/model/config/features/Metric.kt
package io.oddworks.device.model.config.features

import io.oddworks.device.metric.*

class Metric(val type: MetricType, val enabled: Boolean, action: String?, val interval: Int = 0) {

    val action: String

    init {
        if (action == null) {
            this.action = type.defaultAction
        } else {
            this.action = action
        }
    }

    enum class MetricType(val key: String, val defaultAction: String) {
        APP_INIT("appInit", "app:init"),
        VIEW_LOAD("viewLoad", "view:load"),
        VIDEO_LOAD("videoLoad", "video:load"),
        VIDEO_PLAY("videoPlay", "video:play"),
        VIDEO_PLAYING("videoPlaying", "video:playing"),
        VIDEO_STOP("videoStop", "video:stop"),
        VIDEO_ERROR("videoError", "video:error"),
        USER_NEW("userNew", "user:new")
    }

    companion object {
        fun setupOddMetrics(metrics: Set<Metric>) {
            metrics.forEach {
                setupOddMetric(it)
            }
        }

        private fun setupOddMetric(metric: Metric) {
            when(metric.type) {
                MetricType.APP_INIT -> {
                    OddAppInitMetric.action = metric.action
                }
                MetricType.VIEW_LOAD -> {
                    OddViewLoadMetric.enabled = metric.enabled
                    OddViewLoadMetric.action = metric.action
                }
                MetricType.VIDEO_LOAD -> {
                    OddVideoLoadMetric.enabled = metric.enabled
                    OddVideoLoadMetric.action = metric.action
                }
                MetricType.VIDEO_PLAY -> {
                    OddVideoPlayMetric.enabled = metric.enabled
                    OddVideoPlayMetric.action = metric.action
                }
                MetricType.VIDEO_PLAYING -> {
                    OddVideoPlayingMetric.enabled = metric.enabled
                    OddVideoPlayingMetric.action = metric.action
                    OddVideoPlayingMetric.interval = metric.interval
                }
                MetricType.VIDEO_STOP -> {
                    OddVideoStopMetric.enabled = metric.enabled
                    OddVideoStopMetric.action = metric.action
                }
                MetricType.VIDEO_ERROR -> {
                    OddVideoErrorMetric.enabled = metric.enabled
                    OddVideoErrorMetric.action = metric.action
                }
                MetricType.USER_NEW -> {
                    OddUserNewMetric.enabled = metric.enabled
                    OddUserNewMetric.action = metric.action
                }
            }
        }
    }
}

